[v1] Add Whisper model support (encoder-decoder)

russellb opened the pull request:
v1: Add Whisper encoder-decoder model support

Implements Whisper mdoel support in the V1 engine. Key changes include:

- Add encoder-decoder architecture support with cross-attention KV cache management
- Add CrossAttentionManager and CrossAttentionSpec for encoder-decoder KV cache
- Update scheduler to handle cross-attention block allocation and disable prefix caching
- Modify GPU model runner for encoder input processing and attention metadata
- Disable BART tests/examples (Whisper-only support for now)

This closes a major feature gap between V0 and V1, enabling Whisper transcription
in the new engine architecture while maintaining backward compatibility.

Related to V0 deprecation (#18571) and 2025 Q3 roadmap (#20336).

Closes #12761

Signed-off-by: Russell Bryant <rbryant@redhat.com>
Co-authored-by: NickLucche <nlucches@redhat.com>

-----------------

Follow-up TODO items

- [ ] utilize encoder cache
- [ ] remove custom cross-attention slot mapping calculation and integrate into the proper abstraction
- [ ] clean up hard coded assumptions about whisper / multi-modal

russellb committed:
v1: Add Whisper encoder-decoder model support

Implements Whisper mdoel support in the V1 engine. Key changes include:

- Add encoder-decoder architecture support with cross-attention KV cache management
- Add CrossAttentionManager and CrossAttentionSpec for encoder-decoder KV cache
- Update scheduler to handle cross-attention block allocation and disable prefix caching
- Modify GPU model runner for encoder input processing and attention metadata
- Disable BART / other enc-dec tests/examples (Whisper-only support for now)
- Optimize test performance and fix various integration issues

This closes a major feature gap between V0 and V1, enabling Whisper transcription
in the new engine architecture while maintaining backward compatibility.

Related to V0 deprecation (#18571) and 2025 Q3 roadmap (#20336).

Signed-off-by: Russell Bryant <rbryant@redhat.com>
Co-authored-by: NickLucche <nlucches@redhat.com>
Signed-off-by: Russell Bryant <rbryant@redhat.com>

DarkLight1337 commented:
This is already some work to support encoder-decoder models:
- #19988
- #20226

Can you coordinate with @maxdebayser to avoid duplicate work?

maxdebayser commented:
Yeah, I've been talking with @russellb as there are a few overlapping points in our PRs for example disabling prefix caching and chunked prefill. 
Currently in my PR I'm not disabling the KV cache entirely because functionally it makes no difference for the encoder attention. So I can keep the diff small. But I do want to test if removing the KV cache will have a performance improvement for encoder models

russellb commented on vllm/v1/attention/backends/flash_attn.py:
```diff
@@ -552,7 +622,7 @@ def forward(
                 max_seqlen_k=max_seqlen_k,
                 softmax_scale=self.scale,
-                causal=True,
+                causal=FlashAttentionImpl._get_causal_option(attn_type),
```
yes! thanks for the feedback

russellb commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -2740,3 +2874,121 @@ def _maybe_pad_mamba_page_size(
+                            cross_attn_group_idx = i
+                            break
+
+                    if ((cross_attn_group_idx is not None) and
+                        (cross_attn_group_idx < len(req_state.block_ids))):
+                        # Get cross attention block IDs for this request
+                        cross_block_ids = req_state.block_ids[
+                            cross_attn_group_idx]
+                        block_size = self.kv_cache_config.kv_cache_groups[
+                            cross_attn_group_idx].kv_cache_spec.block_size
+
+                        # Calculate slot mapping from block IDs
+                        for i in range(encoder_seq_len):
+                            block_number = cross_block_ids[i // block_size]
+                            block_offset = i % block_size
+                            slot = block_number * block_size + block_offset
+                            cross_slot_mapping.append(slot)
```
The trigger to allocate blocks for cross-attention is when encoder inputs have been scheduled, so I don't think the condition called out here can happen.

NickLucche reviewed (changes requested):
nice one!

russellb commented:
I got this caught up with `main` with all conflicts resolved, but I haven't addressed feedback received so far.

maxdebayser commented on vllm/v1/attention/backends/flash_attn.py:
```diff
@@ -428,6 +470,14 @@ def forward(
+                          AttentionType.ENCODER_ONLY)
+                and (not attn_metadata.is_all_encoder_attn_metadata_set)):
+            raise AttributeError("Encoder attention requires setting "
+                                 "encoder metadata attributes.")
```
ENCODER_ONLY attention doesn't need this, can use the same attributes as DECODER.

maxdebayser commented on vllm/v1/attention/backends/flash_attn.py:
```diff
@@ -477,7 +545,7 @@ def forward(
             scheduler_metadata = attn_metadata.scheduler_metadata
 
-            descale_shape = (cu_seqlens_q.shape[0] - 1, key.shape[1])
+            descale_shape = (cu_seqlens_q.shape[0] - 1, self.num_kv_heads)
```
I was wondering why this change is required.

robertgshaw2-redhat replied:
it might be worth adding this comment around why the `key is not None and value is not None` is needed above because this clarified a lot for me

WoosukKwon reviewed (commented):
I think we haven't made a concrete decision on whether to support the model in V1. Let's discuss offline.

russellb commented:
> I think we haven't made a concrete decision on whether to support the model in V1. Let's discuss offline.

Understood! I worked on this partly to answer the question of how invasive it would be, so we'd have something concrete to evaluate. I'd love to walk through it sometime if we can sync live!

Some key takeaways for me:

- The hybrid memory allocator fit pretty nicely for allocating KV cache blocks for cross attention (ENCODER_DECODER). That alone minimized how invasive this one by a good bit. I did have to make some tweaks to the KV cache manager since ENCODER_DECODER needs a single static allocation instead of an allocation that grows over time, so the logic is a bit different (simpler, though).

- The attention backend changes are bigger for encoder self-attention compared to cross-attention. The encoder side has been split out into #21270.

- Prefix caching is off for whisper (encoder-decoder in general). It's at least a very small and simple change. I do think it's possible to support prefix caching, but I think it's going to make the code a lot more complex than it's worth. The complex part is that hashes can't just be based on the decoder sequence. It must also include a hash of the encoder input/output. Since it isn't then embedded in the decoder sequence (like a LLaVA style model), taking the encoder data into account will require special new handling that's only relevant to encoder-decoder models. I just didn't feel like that complexity was worth it.

russellb commented:
rebased on `main`.

Next step: incorporate changes from #22018 - progress made by @NickLucche while I was out (thanks!)

maxdebayser commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -948,7 +985,8 @@ def _free_encoder_inputs(self, request: Request) -> None:
             start_pos = mm_positions.offset
             num_tokens = mm_positions.length
-            if start_pos + num_tokens <= request.num_computed_tokens:
+            if (self.encoder_cache_manager
```
Is it necessary to test `self.encoder_cache_manager` here again since it's already tested at the top of the function?

russellb replied:
good catch!

russellb commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -948,7 +985,8 @@ def _free_encoder_inputs(self, request: Request) -> None:
             start_pos = mm_positions.offset
             num_tokens = mm_positions.length
-            if start_pos + num_tokens <= request.num_computed_tokens:
+            if (self.encoder_cache_manager
+                    and start_pos + num_tokens <= request.num_computed_tokens):
```
```suggestion
            if start_pos + num_tokens <= request.num_computed_tokens:
```

russellb commented:
> Hello, and thank you for all your great work on this project! ﻿ I'm a user testing the new whisper branch in your v1 implementation and I've observed that the processing time is significantly longer than in v0. After some profiling, it appears that the encode part of the process is being repeatedly computed for a single request. This means that each time a decoding step occurs, the encode result is re-calculated, which seems to be causing a major performance degradation. ﻿ Is this the expected behavior, or could I be using the implementation incorrectly? Any guidance or insight you can provide would be greatly appreciated. Thank you again for your time and effort!

Thank you for catching that!

I know it definitely did NOT do that earlier, but there has been quite a lot of rebasing and resolving merge conflicts, so I'm not surprised that a bug snuck in. I appreciate you doing some profiling to find the cause!

I will get that fixed and update the PR when I sort it out.

russellb commented:
@Sugar-zsg the performance issue should now be resolved. Thanks again for testing this.

I still need to update it again to resolve conflicts with #21588 before it's fully up to date with `main` again, but it should work for testing the performance if you'd like to check again.

russellb commented:
and now it's up to date with `main` again and ready for review

NickLucche commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -131,14 +132,19 @@ def __init__(
+            # An encoder-decoder model does not use the encoder cache.
+            # It uses bidirectional attention and inputs are only
+            # processed once per request.
+            self.encoder_cache_manager = EncoderCacheManager(
+                cache_size=encoder_cache_size)
```
I wonder whether a way to keep scheduler cleaner would be passing this `is_encoder_decoder` flag to the cache which makes all methods no-ops and returns `free_encoder_input_ids->[ ]`..

russellb replied:
Yeah, that sounds nicer. I'll give that a try.

russellb replied:
done!

NickLucche commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -788,16 +795,24 @@ def _prepare_inputs(
-            common_attn_metadata, encoder_attn_metadata = \
-                self._build_encoder_only_attn_metadata(
-                scheduler_output)
+        if self.is_encoder_only_model or (
+                self.model_config.is_encoder_decoder
+                and scheduler_output.scheduled_encoder_inputs):
+            if self.is_encoder_only_model:
+                common_attn_metadata, encoder_attn_metadata = \
+                    self._build_encoder_only_attn_metadata(
+                    scheduler_output)
+            else:
+                common_attn_metadata, encoder_attn_metadata = \
+                    self._build_enc_dec_attn_metadata(
+                    scheduler_output)
```
I think we can simplify to

```suggestion
        if self.is_encoder_only_model:
            common_attn_metadata, encoder_attn_metadata = \
                    self._build_encoder_only_attn_metadata(
                    scheduler_output)
        elif self.model_config.is_encoder_decoder and scheduler_output.scheduled_encoder_inputs:
             common_attn_metadata, encoder_attn_metadata = \
                    self._build_enc_dec_attn_metadata(
                    scheduler_output)
```

NickLucche commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -828,6 +843,12 @@ def _prepare_inputs(
                 causal=True,
             )
 
+            is_enc_dec = isinstance(kv_cache_group_spec.kv_cache_spec,
```
don't we know this is an enc_dec model already?

russellb replied:
We do, but here we need to know if we're looking at a cross-attention layer or regular self-attention. So the check is about attention type, not the model type. We use `AttentionType.ENCODER_DECODER` so the attention type and model type share the same name in our code ... it's a little confusing sometimes.

NickLucche commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -1212,6 +1234,89 @@ def _gather_mm_embeddings(
+                        total_encoder_tokens += num_features * \
+                            self.max_encoder_len
+
+        if not input_features_list:
+            return {}
```
nit: I don't think text only models like bart would require `input_features`, yet they may still need `encoder_positions`..

NickLucche commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -1212,6 +1234,89 @@ def _gather_mm_embeddings(
+
+        return {
+            "input_features": input_features,
+            "encoder_input_ids": encoder_input_ids,
+            "encoder_positions": encoder_positions,
```
I don't think we need them for whisper, and given `encoder_input_ids` is just a mock, perhaps we can get rid of them and leave a TODO for other models?

russellb replied:
i'll look at dropping them

NickLucche commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -1212,6 +1234,89 @@ def _gather_mm_embeddings(
                 mm_embeds.append(mm_embeds_item)
         return mm_embeds
 
+    def _extract_encoder_inputs(
+        self,
+        scheduler_output: "SchedulerOutput",
+    ) -> dict[str, torch.Tensor]:
```
I think this method and the one below would best fit within the `_prepare_inputs` context, so that we could think about re-using a singly allocated on-device tensor for both encoder_inputs.. and input_features like it happens for the rest of params in V1.

We can also leave this as future optimization work.

russellb replied:
Thanks for the feedback. Let me know if you think we need to block on it.

NickLucche reviewed (changes requested):
Left a few comments, the main one being about the possibility of unifying enc-dec forward with MM forward (based on `inputs_embeds` ).

Rest of the changes look clean to me, in particular the kv blocks handling fits well in the current hma framework.


re @Sugar-zsg 
> I noticed that CUDA graphs are not being utilized, even though the relevant parameters are enabled

I think the model is not torch compiled yet, so no splitwise cuda graph is produced either.

russellb commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -1479,14 +1584,16 @@ def execute_model(
         # _prepare_inputs may reorder the batch, so we must gather multi
         # modal outputs after that to ensure the correct order
-        if self.is_multimodal_model:
+        if (self.is_multimodal_model
+                and not self.model_config.is_encoder_decoder):
```
It's not trivial, but possible. 

I went this route thinking that the result would be simpler. Whisper is simpler in some ways in that we have no need to touch the encoder output. I liked the idea of just calling the model forward() and skipping the multimodal code in the runner that doesn't apply to whisper.

I never fully implemented the alternative that you're suggesting, though. That way was my first attempt. I had some trouble and ended up going this route.

If you feel strongly about this, please let me know and I can work on it, but I don't think it'd be a quick change.

russellb replied:
I'm going to try to do this after all to simplify the PR. There are some other open comments about it as well, so I'm going to resolve this one, but wanted to note for the record that I'm going to address it and appreciate you being one of the people calling out that I should try to do it.

russellb reviewed (commented):
thanks for the review!

joennlae commented on vllm/config.py:
```diff
@@ -4872,22 +4873,40 @@ def __post_init__(self):
                 disable_chunked_prefill_reasons.append(
-                    "Only \"last\" pooling supports chunked "
-                    "prefill and prefix caching; disabling both.")
+                    "Encoder-decoder models do not support chunked prefill nor"
```
Very nice PR :-)

Is there a chance that prefix caching will be supported? Or is the effort unreasonable?

joennlae commented on vllm/v1/core/single_type_kv_cache_manager.py:
```diff
@@ -560,11 +561,62 @@ def allocate_new_blocks(self, request_id: str,
+        #    inputs)
+        # 2. Encoder states are computed once per request, not incrementally
+        # 3. No reusable prefix exists between different multimodal inputs
+        # Return empty blocks to indicate no cache hits
```
I’m exploring a “budget” beam search that leverages a prefix cache so we don’t have to change core components to get "reasonable" beam-search performance.

Prototype works with `voxtral` in our fork: https://github.com/vllm-project/vllm/compare/main...44ai-labs:vllm:audio-beam-search

The same audio input is reused across multiple requests (in the beams).

Would adding prefix-cache support for this “budget” beam search be feasible in upstream vLLM? Or is that an unreasonable effort?

heheda12345 commented on vllm/v1/core/encoder_cache_manager.py:
```diff
@@ -99,6 +106,9 @@ def allocate(self, request: Request, input_id: int) -> None:
             This method assumes can_allocate() returned True for the same
             request and input_id. It will reduce available cache space.
         """
+        if self.is_encoder_decoder:
```
should we always return true for `can_allocate`?

heheda12345 commented on vllm/v1/core/single_type_kv_cache_manager.py:
```diff
@@ -560,11 +561,62 @@ def allocate_new_blocks(self, request_id: str,
+        assert isinstance(kv_cache_spec, CrossAttentionSpec), (
+            "CrossAttentionManager can only be used for cross-attention groups"
+        )
+        # Cross-attention does not benefit from prefix caching since:
```
Should we just disable prefix caching globally for encoder-decoder models? Then we can just raise an error if this function is called.

heheda12345 reviewed (commented):
Two major concerns:
1. KVCacheManager: can we hide the details of encoder kv cache vs decoder kv cache, and make the scheduler call the kv cache manager only once?
2. Attention metadata builder: we are trying to make all attribute that not required by standard full attention pluggable. Please see some of the examples in my comments.

Can you also update the document?

robertgshaw2-redhat commented on vllm/v1/attention/backends/flash_attn.py:
```diff
@@ -469,7 +482,7 @@ def forward(
 
         # Handle encoder attention differently - no KV cache needed
-        if attn_type in (AttentionType.ENCODER_ONLY, ):
+        if attn_type in (AttentionType.ENCODER_ONLY, AttentionType.ENCODER):
```
Do we need both `ENCODER_ONLY` and `ENCODER`

russellb replied:
It's a great question. I never found a case where they were different when it comes to AttentionType. It seems to be confusing the type of model with type of attention?

Combining them would be a good follow-up

robertgshaw2-redhat commented on vllm/v1/core/encoder_cache_manager.py:
```diff
@@ -48,14 +49,20 @@ class EncoderCacheManager:
     """
 
-    def __init__(self, cache_size: int):
+    def __init__(self, cache_size: int, is_encoder_decoder: bool):
```
I am still digging in, but why is this special case needed? Shouldn't we be able to re-use this encoder cache? I understand its not explicitly needed since we must schedule all the encoder tokens at once, but perhaps this would simplify the implementation and branching if we did not have this if...else

russellb replied:
The encoder cache is used for referencing encoder output potentially more than once during model execution. Whisper is a bit simpler, where the encoder output is processed only once in all cases.

There may be value in caching encoder output for use across requests, but that's not the same logic implemented in the encoder cache right now, and I decided that was best left as a follow-up optimization.

I looked at using the encoder cache anyway in case it helped clean up the code to reduce special cases, but it really didn't. The processing just isn't the same as the other multimodal models.

robertgshaw2-redhat commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -572,7 +596,7 @@ def schedule(self) -> SchedulerOutput:
             # the previous and the current steps.
             finished_req_ids=self.finished_req_ids,
-            free_encoder_input_ids=self.encoder_cache_manager.get_freed_ids(),
+            free_encoder_input_ids=free_encoder_input_ids,
```
nit - unnecessary change

robertgshaw2-redhat commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -464,9 +484,12 @@ def schedule(self) -> SchedulerOutput:
                 # This information is used to determine if a load is
                 # needed for this request.
                 if self.connector is not None:
+                    update_blocks = new_computed_blocks + new_blocks
```
I think it is better for us to disable KVConnector + Encoder/Decoder for this PR. We have not really tested that it works so I would rather we fail for now

robertgshaw2-redhat commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -436,6 +439,23 @@ def schedule(self) -> SchedulerOutput:
+                            # encoder length. This is a single static allocation
+                            # and does not grow with the number of decoder
+                            # tokens.
+                            max_encoder_len = MULTIMODAL_REGISTRY.\
```
shouldn't this value be dynamic based on the actual input size for the encoder branch?

e.g. for Whisper, if we get a 5s clip rather than 30s clip, do we need to allocate for the full 30s?

cc @NickLucche

robertgshaw2-redhat commented on vllm/config/__init__.py:
```diff
@@ -3838,22 +3839,40 @@ def __post_init__(self):
                 disable_chunked_prefill_reasons.append(
-                    "Only \"last\" pooling supports chunked "
-                    "prefill and prefix caching; disabling both.")
+                    "Encoder-decoder models do not support chunked prefill nor"
```
I dont actually see where we disable prefix caching

robertgshaw2-redhat reviewed (commented):
I think that this PR is looking good

The primary item I would suggest changing is that we should try to use the Embedding Cache for the embeddings coming from the Encoder branch. Its actually not necessary to use the cache (as you have here), but this will avoid a signficant number of `if...else` branches and special casing for `is_encoder_decoder` in the model runner which will help with overall maintenance and code readability over time

robertgshaw2-redhat reviewed (commented):
Looking at the implementation in more detail, I think that the key blocker to using the Embedding cache is that we currently use the `Attention` class for the encoder branch self-attn, which means that we need to pass `AttentionMetadata`, which means that we need to run the Encoder branch inside the `ForwardContext`

Instead, I think we should avoid using the `Attention` class for the Encoder branch and instead just follow how the encoder branch in the MM models do it, i.e. using SPDA direclty
- https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/pixtral.py#L824-L834

We should follow up this PR by making the same change to https://github.com/vllm-project/vllm/pull/21270/files#diff-9621929f76e8de54a4e48af911cc156f7519e60d2faec1809e904752e0844ae7

- Another thought is that we actually ** could ** move the CROSS-ATTENTION logic "fill" of the KV cache outside of the forward pass of the decoder branch (e.g. just the reshape and cache). If you look at Whisper (https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/whisper.py#L270-L283), you can see the computation of the K and V for CROSS-ATTN only depends only on the encoder hidden states --- this might be more trouble than its worth, but it would allow us to avoid passing the `slot_mapping_cross` via the existing AttentionMetadata objects

```python
# Encoder hidden states are only computed once during prefill phase.
# Afterwards, the keys and values should be available in the kv-cache.
if encoder_hidden_states is not None:
    kv, _ = self.kv_proj(encoder_hidden_states)
    k, v = kv.split([self.kv_size, self.kv_size], dim=-1)
```

robertgshaw2-redhat commented on vllm/v1/attention/backends/utils.py:
```diff
@@ -64,6 +64,14 @@ class CommonAttentionMetadata:
     causal: bool = True
 
+    # Encoder/cross-attention specific fields (optional)
+    encoder_seq_start_loc: Optional[torch.Tensor] = None
```
I don't think this is being used anywhere?

russellb replied:
removed!

LucasWilkinson commented on vllm/v1/attention/backends/flash_attn.py:
```diff
@@ -220,7 +229,13 @@ def build(self,
+            # ENCODER_DECODER cross-attention
+            max_seq_len = common_attn_metadata.max_encoder_seq_len
+        else:
+            max_seq_len = int(common_attn_metadata.seq_lens_cpu.max())
```
in the encoder case is `max_encoder_seq_len` not equal to `int(common_attn_metadata.seq_lens_cpu.max())`?

LucasWilkinson replied:
if we can get rid of this and `updated_slot_mapping = attn_metadata.cross_slot_mapping` we reduce the amount of changes other backends would have to make to support encoder-decoder

russellb commented:
last push was fixing multiple conflicts again -- wanted to do that before working on review feedback

russellb commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -946,6 +981,8 @@ def _update_request_with_output(
         return new_token_ids, stopped
 
     def _free_encoder_inputs(self, request: Request) -> None:
+        if not self.encoder_cache_manager:
```
it's not. it is left over from a previous change where I had disabled the encoder cache manager by not creating it. Later, I just changed its behavior internally to get rid of checks like this. I just missed this one.

russellb commented on vllm/v1/attention/backends/flash_attn.py:
```diff
@@ -481,20 +494,26 @@ def forward(
             # op uses the slot_mapping's shape to determine the number of
             # actual tokens.
+            if attn_type == AttentionType.ENCODER_DECODER:
+                updated_slot_mapping = attn_metadata.cross_slot_mapping
```
done -- all of the encoder related fields have been removed from common attn metadata in favor of using existing fields! yay

russellb commented on vllm/v1/attention/backends/flash_attn.py:
```diff
@@ -139,6 +139,15 @@ class FlashAttentionMetadata:
+    encoder_seq_start_loc: Optional[torch.Tensor] = None
+    # Maximum sequence length among encoder sequences
+    max_encoder_seq_len: Optional[int] = None
+    cross_slot_mapping: Optional[torch.Tensor] = None
```
I have removed all of these encoder specific fields in favor of re-using existing attributes

russellb commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -436,6 +439,23 @@ def schedule(self) -> SchedulerOutput:
+                            max_encoder_len = MULTIMODAL_REGISTRY.\
+                                get_encdec_max_encoder_len(
+                                self.vllm_config.model_config)
+                            new_cross_blocks = (self.kv_cache_manager.
```
They're included later when this is called: 

```python
                req_to_new_block_ids[request.request_id] = (
                    self.kv_cache_manager.get_block_ids(request.request_id))
```

This will include two lists of block IDs. The first is associated with decoder self-attention, and the second is for decoder cross-attention.

russellb commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -436,6 +439,23 @@ def schedule(self) -> SchedulerOutput:
+                                                    max_encoder_len,
+                                                ))
+                            if new_cross_blocks is None:
+                                # The request cannot be scheduled.
```
Good catch!

Right now the cross-attention blocks are allocated first. If the decoder-only attention allocation fails here, we will not release the cross-attention allocation. That allocation is going to be held until this request is able to finish allocating all of the blocks it needs to run.

I suppose the better behavior would be to release the blocks that were allocated in case they're needed for something else and only keep them if we know the request is getting scheduled in this batch.

I'll fix this to release the cross attention allocation.

russellb commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -714,7 +738,18 @@ def _try_schedule_encoder_inputs(
                 break
-            if start_pos + num_encoder_tokens <= num_computed_tokens:
+            if self.is_encoder_decoder and start_pos < num_computed_tokens:
+                # Encoder input has already been computed
```
Yeah, I like that -- I'd want to know if start_pos != 0, since then something would be breaking my assumption here

russellb commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -827,14 +834,20 @@ def _prepare_inputs(
-                self._build_encoder_only_attn_metadata(
-                scheduler_output)
+                    self._build_encoder_only_attn_metadata(
+                    scheduler_output)
```
It looks like it's still there about 10 lines down?

heheda12345 replied:
Yes you are right

russellb commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -827,14 +834,20 @@ def _prepare_inputs(
+        elif (self.model_config.is_encoder_decoder
+              and scheduler_output.scheduled_encoder_inputs):
+            common_attn_metadata, encoder_attn_metadata = \
+                    self._build_enc_dec_attn_metadata(
```
Yes, you're right. I'll remove this code from here.

russellb replied:
Well, if I use the same metadata that was created for the encoder_only model code path, it fails for whisper. It does seem like we should be able to reduce duplication here, but I'll have to revisit it later.

russellb replied:
I revisited this and was able to better unify these additions with the metadata building for 
encoder-only model support. There's one method for building metadata for encoder attention (whether it's an encoder-only or encoder-decoder model), and another for cross-attention.

russellb commented on vllm/v1/worker/utils.py:
```diff
@@ -277,14 +277,14 @@ def bind_kv_cache(
-        layer_name = layer_names[0]
-        runner_kv_caches.append(kv_caches[layer_name])
+        for layer_name in layer_names:
+            runner_kv_caches.append(kv_caches[layer_name])
```
Thanks for the suggestion. I have implemented your suggestion here. I reverted the change, made `runner_kv_caches` optional, and udpated the GPU runner to pass None here.

russellb replied:
I did, and then I reverted it and forgot to update this thread.

Allowing `runner_kv_caches=None` for the GPU model runner broke a test that depended on it.

I think it might be safe to just revert these changes completely, except we can't raise `NotImplementedError`.

russellb replied:
Since I know leaving the code as-is is safe for the GPU runner, I'm going to skip the exception for `current_platform.is_cuda()` and leave a TODO to analyze further for other device types.

russellb committed:
whisper: simplify encoder attention by using pytorch

We don't need the extra features of the attention backends available
through the more involved Attention abstraction. Replace encoder
attention with a new `TorchAttention` class which uses
scaled_dot_product_attention.

This also simplifies how the encoder can be run, as it no longer needs
to be run within a forward context with all of the attention metadata
available.

Signed-off-by: Russell Bryant <rbryant@redhat.com>

heheda12345 reviewed (commented):
If cross attention only supports flash attention backend, can you add an assert on it?
The `KVCacheManager` related code is quite clean now. I only left some small comments.

But I'm a little bit concern about `_build_enc_dec_attn_metadata` and still thinking of how can we get a cleaner implementation. Also CC @LucasWilkinson for this concern.

russellb commented on vllm/v1/attention/backends/flash_attn.py:
```diff
@@ -489,20 +483,23 @@ def forward(
 
-        if self.kv_sharing_target_layer_name is None:
+        if (self.kv_sharing_target_layer_name is None and (key is not None)
+                and (value is not None)):
```
Yes, I'll add a comment. It would be something like this:

> `key` and `value` may be `None` in the case of cross attention. They are calculated once based on the output from the encoder and then cached in KV cache.

russellb commented on vllm/v1/attention/backends/flash_attn.py:
```diff
@@ -489,20 +483,23 @@ def forward(
                 key_cache,
                 value_cache,
-                attn_metadata.slot_mapping,
+                updated_slot_mapping,
```
It's not needed anymore. I used to have a separate `cross_slot_mapping` attribute in the attention metadata, but it has since been removed. This is left over from when I set `updated_slot_mapping` based on the attention type. I will remove this.

russellb commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -445,13 +450,26 @@ def schedule(self) -> SchedulerOutput:
+                    # NOTE(russellb): For Whisper, we know that the input is
+                    # always padded to the maximum length. If we support other
+                    # encoder-decoder models, this will need to be updated if we
+                    # want to only allocate what is needed.
```
Yes, I will add the assertion.

I'm going to work on BART support next, so it will probably be me fixing this very soon.

russellb commented on vllm/v1/kv_cache_interface.py:
```diff
@@ -203,6 +204,24 @@ def max_memory_usage_bytes(self, vllm_config: VllmConfig) -> int:
+    """
+
+    @property
+    def type_id(self) -> str:
```
Thank you, I will update this

russellb commented:
I split out another chunk from this PR here: https://github.com/vllm-project/vllm/pull/23664

This PR has major conflicts that I'm working on this week, mostly from #23154

russellb commented:
Updated to fix more conflicts.

It's working when I test manually, but one of the tests is triggering a memory error. I'm trying to debug that.

russellb committed:
prevent voxtral from being detected as encoder-decoder

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
Drop whisper-small from tests

small is less reliable and doesn't seem to add much value in making
sure whisper is working.

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
Use correct number of encoder tokens in attention metadata

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -815,26 +831,125 @@ def _prepare_inputs(
+                    _dummy_blk_table_and_slot_mapping())
+                num_common_prefix_blocks = 0
+                causal_arg = False
+            elif isinstance(kv_cache_group_spec.kv_cache_spec,
```
Right - I had this in my notes here https://github.com/vllm-project/vllm/pull/21088#issuecomment-3234693048

I just hadn't gotten to it, since it took me a bit to get it all working again. I'm also fixing conflicts almost every day.

RIght now I'm focused on going through CI failures. I'll keep this on my TODO list.

russellb committed:
Force spawn multiproc method in whisper example

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
Warn if Whisper is used without spawn

It causes startup to hang for me with this warning:

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
        - Avoid using `tokenizers` before the fork if possible
        - Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
Run whisper test with spawn multiproc method

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
Limit whisper encoder concurrency

Allow the encoder cache manager in the scheduler to limit how how many
encoder inputs we are processing in a single run of the model. We were
previously allowing too many at once, leading to out-of-memory failures.

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
Drop EncoderAttention abstractions no longer needed

Since we're using attention from pytorch directly and not using our
attention abstraction to multiple backends, this gets much simpler.

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
remove debug logs

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb commented:
Status:

- simplified encoder attention to just use PyTorch instead of attention backends (see `TorchAttention`). This removed a lot of code related to encoder attention metadata.
- fix code inside `profile_run()` to run the Whisper encoder for a more accurate memory consumption profile
- fix scheduler not to allow an unbounded number of whisper encoder inputs to be processed at the same time

I don't see any OOM failures in CI. Some failures look unrelated. There is one that looks like an accuracy issue that I'm trying to track down.

russellb committed:
encoder-decoder does not use encoder-cache yet

This fixes cases where we would skip running the encoder completely for
some requests when sent in parallel.

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
Simplify and reduce duplication in TorchAttention

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb commented:
This is ready for another review. The current CI failures appear to be unrelated.

Here is my current list of follow-up items. I don't consider them blockers unless reviewers feel otherwise. I will file issues for all of my follow-up items and continue working on them.

- The manual calculation of the cross-attention slot mappings is not ideal. This needs to be integrated into the proper abstraction.
- The encoder cache was recently updated to cache inputs across different requests, making it useful for Whisper, but I'm not utilizing it yet.

heheda12345 commented on vllm/attention/layers/cross_attention.py:
```diff
@@ -0,0 +1,185 @@
+    # Get cross attention block IDs and calculate slot mapping
+    cross_block_ids = req_state.block_ids[cross_attn_group_idx]
+    block_size = kv_cache_config.kv_cache_groups[
+        cross_attn_group_idx].kv_cache_spec.block_size
```
you can get `kv_cache_spec` from `underlying_builder.kv_cache_spec`

russellb committed:
Use existing MultiHeadAttention instead of new TorchAttention

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb commented on vllm/attention/layers/cross_attention.py:
```diff
@@ -0,0 +1,185 @@
+            break
+
+    if (cross_attn_group_idx is None
+            or cross_attn_group_idx >= len(req_state.block_ids)):
```
no, this should be removed

russellb commented on vllm/attention/layers/cross_attention.py:
```diff
@@ -0,0 +1,185 @@
+        seq_lens=seq_lens_arg,
+        seq_lens_cpu=seq_lens_cpu_arg,
+        num_computed_tokens_cpu=common_attn_metadata.num_computed_tokens_cpu,
+        num_reqs=num_reqs,
```
done, thanks!

russellb commented on vllm/attention/layers/cross_attention.py:
```diff
@@ -0,0 +1,185 @@
+    req_state = requests.get(req_id)
+    if req_state is None:
+        # During memory profiling or if request not found
+        return [PAD_SLOT_ID] * encoder_seq_len
```
done, thanks!

russellb committed:
Remove some unnecessary variables that were not used

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb commented:
@heheda12345 I pushed an update. I reverted all of the additions to CommonAttentionMetadata. It's back to doing more in the GPU model runner, but I couldn't come up with a clean solution pushing more of it into the new cross-attn metadata builder. I also had to spend a bunch of time resolving conflicts again. It would be super helpful if we could merge this as it is and continue working on it in main.

russellb committed:
Remove unused TorchAttention

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
Replace a slow Python loop with torch

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb commented on vllm/attention/layers/cross_attention.py:
```diff
@@ -0,0 +1,185 @@
+        cross_attn_group_idx].kv_cache_spec.block_size
+
+    slot_mapping = []
+    for i in range(encoder_seq_len):
```
done, thanks

russellb committed:
Move max_seq_len override into CrossAttentionBuilder

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
Move seq_lens / seq_lens_cpu overrides into CrossAttentionBuilder

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
Revert unnecessary code move to reduce diff size

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
further simplification of slot_mappings for cross attn

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
move slot mapping calculation back into CrossAttentionBuilder

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
improve how we get the number of blocks needed

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
make python loop more efficient

Signed-off-by: Russell Bryant <rbryant@redhat.com>

heheda12345 commented on vllm/attention/layer.py:
```diff
@@ -379,6 +379,12 @@ def forward(
         value: torch.Tensor,
     ) -> torch.Tensor:
         """Input shape: batch_size x seq_len x hidden_size"""
+        is_2d = query.dim() == 2
```
Can you do it in model code?

heheda12345 commented on vllm/attention/layers/cross_attention.py:
```diff
@@ -0,0 +1,160 @@
+        encoder_seq_len = encoder_seq_lens[req_index]
+
+        # Calculate the number of blocks needed for this request
+        num_blocks_needed = (encoder_seq_len + block_size -
+                             1) // block_size  # Ceiling division
```
```suggestion
        num_blocks_needed = cdiv(encoder_seq_len, block_size)
```
with `from vllm.utils import cdiv`

russellb replied:
done, thanks!

heheda12345 commented on vllm/attention/layers/cross_attention.py:
```diff
@@ -0,0 +1,160 @@
+                  fast_build: bool = False) -> AttentionMetadata:
+            new_metadata = copy(common_attn_metadata)
+            new_metadata.causal = False
+            max_encoder_len = _get_max_encoder_len(self.vllm_config)
```
Can we use `max(common_attn_metadata.encoder_seq_lens)` now?

heheda12345 commented on vllm/attention/layers/cross_attention.py:
```diff
@@ -0,0 +1,160 @@
+            max_encoder_len = _get_max_encoder_len(self.vllm_config)
+            new_metadata.max_seq_len = max_encoder_len
+
+            new_metadata.seq_lens = torch.full(
```
is it `common_attn_metadata.encoder_seq_lens`?

russellb replied:
See my reply about `            max_encoder_len = _get_max_encoder_len(self.vllm_config)`

heheda12345 commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -367,6 +381,9 @@ def __init__(
         self.runner_only_attn_layers: set[str] = set()
 
+        # Cache for cross-attention group index
+        self._cross_attn_group_idx: Optional[int] = None
```
Do you still need this?

heheda12345 commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -3383,6 +3473,8 @@ def initialize_kv_cache(self, kv_cache_config: KVCacheConfig) -> None:
         kv_cache_config = deepcopy(kv_cache_config)
         self.kv_cache_config = kv_cache_config
+        # Invalidate cached cross-attention group index
+        self._cross_attn_group_idx = None
```
Do you need still this?

russellb replied:
No, good catch, thanks

heheda12345 commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -3466,9 +3561,6 @@ def get_kv_cache_spec(self) -> dict[str, KVCacheSpec]:
                 continue
 
-            # TODO: Support other attention modules, e.g., cross-attention
-            # TODO(lucas): move the attention specs into the model layers like
```
Can you keep the second TODO left by lucas?

russellb replied:
Yes! Sorry about that. I probably removed it by accident while resolving conflicts at some point.

heheda12345 commented on vllm/attention/layer.py:
```diff
@@ -414,7 +420,12 @@ def forward(
-        return out.reshape(bsz, q_len, -1)
+        out = out.reshape(bsz, q_len, -1)
+
+        if is_2d:
```
Can you do it in model code?

russellb replied:
I moved it to the model code, thanks

heheda12345 commented on vllm/v1/attention/backends/flash_attn.py:
```diff
@@ -177,6 +177,7 @@ class FlashAttentionMetadataBuilder(
 
     def __init__(self, kv_cache_spec: AttentionSpec, layer_names: list[str],
                  vllm_config: VllmConfig, device: torch.device):
+        super().__init__(kv_cache_spec, layer_names, vllm_config, device)
```
Do other attention backends need this line? (Can be a follow-up PR)

russellb replied:
Yes, they probably do. I'll check and include the changes here since it should be quick.

heheda12345 commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -955,6 +973,15 @@ def _prepare_inputs(
                     scheduler_output.
                     num_common_prefix_blocks[kv_cache_group_id])
 
+                if isinstance(kv_cache_group_spec.kv_cache_spec,
```
```
        encoder_seq_lens: Optional[np.ndarray] = None
        if isinstance(kv_cache_group_spec.kv_cache_spec, CrossAttentionSpec):
            encoder_seq_lens= self._prepare_encoder_seq_lens(necessary_args)
```
Can you put this part of code into a seperate function like above?

And I slightly prefer to put this under the for loop than to put it under the `if isinstance(EncoderOnlyAttentionSpec): xxx; else xxx` block.

heheda12345 commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -1242,10 +1270,27 @@ def _prepare_kv_sharing_fast_prefill(
+            scheduler_output: The scheduler output containing scheduled encoder
+              inputs.
+
+            include_positions: Whether to also return position information for
```
Do you still need this arg?

heheda12345 commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -3416,6 +3508,9 @@ def initialize_kv_cache(self, kv_cache_config: KVCacheConfig) -> None:
         """
         Add encoder-only layers to the KV cache config.
+        
+        This handles:
```
This docstring is a bit weird to me now.

russellb replied:
Yeah, this is left over from when I had expanded this function. I missed removing all of the docstring changes. Thanks!

heheda12345 reviewed (commented):
Really nice now! Thanks for your effort. I left some small comments.

russellb committed:
remove leftover docstring addition

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
Restore TODO that was accidentally removed

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
remove variables no longer needed

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
Ensure AttentionMetadataBuilder subclasses call parent constructor

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
remove old param from docstring

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
move MultiHeadAttention changes into whisper.py

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
use vllm.utils.cdiv

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
minor refactoring

Signed-off-by: Russell Bryant <rbryant@redhat.com>

heheda12345 commented on vllm/v1/core/encoder_cache_manager.py:
```diff
@@ -166,7 +166,6 @@ def allocate(self, request: Request, input_id: int) -> None:
         Note:
             This method assumes can_allocate() returned True for the same input.
         """
-
```
revert?

heheda12345 reviewed (approved):
LGTM! Thanks for your patience for iterating on this. It is really helpful for v0 deprecation.

russellb committed:
revert unnecessary changes left over

Signed-off-by: Russell Bryant <rbryant@redhat.com>
